app/models/user.rb

```diff
@@ -0,0 +1,33 @@
+class User < ActiveRecord::Base
+  attr_accessor :login
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+
+  validates :username, presence: true, length: {maximum: 255},
+    uniqueness: { case_sensitive: false },
+    format: { with: /\A[a-zA-Z0-9]*\z/,
+    message: "may only contain letters and numbers." }
+
+  validate :validate_username
+
+   def self.find_for_database_authentication(warden_conditions)
+     conditions = warden_conditions.dup
+     if login = conditions.delete(:login)
+       where(conditions.to_hash)
+       .where(["lower(username) = :value OR lower(email) = :value",
+         { :value => login.downcase }]).first
```

Comment:
Align the elements of an array literal if they span more than one line.<br>Use the new Ruby 1.9 hash syntax.